apps/api/src/affiliate/affiliate.controller.ts

```diff
@@ -0,0 +1,114 @@
+import { AffiliateService } from './affiliate.service'
+import { AffiliateStatusUpdateDto } from './dto/affiliate-status-update.dto'
+import { CreateAffiliateDonation } from './dto/create-affiliate-donation.dto'
+import { DonationsService } from '../donations/donations.service'
+import { CancelAffiliateDonation } from './dto/cancel-affiliate-donation.dto'
+import { shouldAllowStatusChange } from '../donations/helpers/donation-status-updates'
+import { affiliateCodeGenerator } from './utils/affiliateCodeGenerator'
+
+@Controller('affiliate')
+@ApiTags('affiliate')
+export class AffiliateController {
+  constructor(
+    private readonly personService: PersonService,
+    private readonly affiliateService: AffiliateService,
+    private readonly donationService: DonationsService,
+  ) {}
+
+  @Post('join')
+  async joinAffiliateProgramRequest(@AuthenticatedUser() user: KeycloakTokenParsed) {
+    const person = await this.personService.findOneByKeycloakId(user.sub)
+    if (!person) throw new NotFoundException('User is not found')
+    if (!person.company) throw new BadRequestException('Must be corporate profile')
+    return await this.affiliateService.create(person.company.id)
+  }
+
+  @Patch(':affiliateId/status-update')
+  async updateAffiliateStatus(
+    @Param('affiliateId') affiliateId: string,
+    @Body() { newStatus }: AffiliateStatusUpdateDto,
+    @AuthenticatedUser() user: KeycloakTokenParsed,
+  ) {
+    if (!isAdmin(user)) throw new ForbiddenException('Must be an admin ')
+    const affiliate = await this.affiliateService.findOneById(affiliateId)
+
+    if (!affiliate) throw new NotFoundException('Affiliate not found')
+
+    let affiliateCode: string | null = affiliate.affiliateCode
+
+    if (affiliate.status === newStatus) {
+      throw new ConflictException('Status is the same')
+    }
+
+    if (affiliate.status !== 'active' && newStatus === 'active') {
+      affiliateCode = affiliateCodeGenerator(affiliate.id)
+    }
+
+    if (affiliate.status === 'active' && newStatus !== 'active') {
+      affiliateCode = null
+    }
+
+    return await this.affiliateService.updateStatus(affiliateId, newStatus, affiliateCode)
+  }
+
+  @Get(':affiliateCode')
+  @Public()
+  async affiliateSummary(@Param('affiliateCode') affilliateCode: string) {
+    return await this.affiliateService.findOneByCode(affilliateCode)
+  }
+
+  @Post(':affiliateCode/donations/create')
+  @Public()
+  async createAffiliateDonation(
+    @Param('affiliateCode') affiliateCode: string,
+    @Body() donation: CreateAffiliateDonation,
+  ) {
+    const affiliate = await this.affiliateService.findOneByCode(affiliateCode)
+    if (!affiliate || !affiliate.company || !affiliate.company.person)
+      throw new NotFoundException('Affiliate not found')
+
+    const affiliateDonationDto: CreateAffiliateDonation = {
+      ...donation,
+      affiliateId: affiliate.id,
+      personId: donation.isAnonymous ? null : affiliate.company.person.id,
```

Comment:
The `affiliate.company.person.id` , represents the company, thus any donations connected to that field, should be considered to be done by the company, rather than the individual person behind the profile, be it affiliate donation or regular donation.

The `personId` of the (end end user), doesn't exist in our database records. We could expand the request body to include(personEmail, personName), and create a new person, but then donation certificate will be issued for that individual person, rather than for the company, which is actually making the transaction. To avoid that we could also expand the affiliate API, to allow the affiliate admin to create more corporate profiles(which don't require manual activation), but then the keycloak should be configured to allow only for the **affiliate administrator** to be able to create such profiles, and it kind of becomes overcomplicated.

The idea was to use the `billingName` field to mark the exact person who made the donation(which I seem to forgot to put inside the object). 
Seems like I also forgot to include the field `extCustomerId` which would allow the affiliate to internally track the profile behind the donation.

Overall marking the actual person behind the affiliate donation is kind of hard. Even if we use `billingName` field to get that person, we need to add additional complexity to the `listPublic` endpoint, to transform the donations results, in a way that `billingName` is only included in the response, if the donation is affiliate, and it isn't anonymous.

If we just include `billingNamel` in the listPublic response, then no donation would actually be anonymous. 

I will include the  the forgotten `billingName` and  `extCustomerId` fields in the createDonationDto, but perhaps the whole process of  getting the exact person behind this affiliate donation could be done  later?